### PR TITLE
fix buildcache

### PR DESCRIPTION
**Pull request by alexvpickering**
# Description
fixes docker buildcache so images dont need full rebuild on every commit.

**not sure if necessary - waiting to see if some fixes from Martin resolve this as well**

**NOTES**: 
- no way AFAIK to tell if it works based on PR (ci.yaml runs as master for PRs)
- tested locally:
  - setup buildx manually: `docker buildx create --use --driver=docker-container`
  - run `docker buildx build ...` commands
  - run `docker buildx build --push ...` command to push buildcache to ECR
  - run `docker push ...` command to push regular image to ECR
  - re-run `docker buildx build ...` command after pruning buildx cache to check that it uses the cache

# Details
#### URL to issue
https://github.com/hms-dbmi-cellenics/issues/issues/21

#### Link to staging deployment URL (or set N/A)
N/A

#### Links to any PRs or resources related to this PR
https://github.com/hms-dbmi-cellenics/pipeline/pull/325

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [ ] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [ ] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [ ] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [ ] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [ ] Started end-to-end tests on the latest commit.

### Documentation updates
- [ ] Relevant Github READMEs updated **or** no GitHub README updates required.
- [ ] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [ ] Photo of a cute animal attached to this PR.

### CODE REVIEW

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -198,29 +198,22 @@ jobs:
...
-        name: Pull latest Docker images from mainline
-        run: |-
-          echo Pulling $IMAGE_NAME from ECR.
-          docker pull $IMAGE_NAME || true
-
-          echo Pulling $IMAGE_NAME-builder from ECR.
-          docker pull $IMAGE_NAME-builder || true
-
-        env:
-          IMAGE_NAME: ${{ format('{0}/{1}:{2}-{3}', steps.login-ecr.outputs.registry, steps.ref.outputs.repo-name, steps.ref-previous.outputs.tag, matrix.project) }}
```
Comment (alexvpickering) [inline .github/workflows/ci.yaml:221]: this is replaced by `--cache-from` argument in `docker buildx build` command

File: .github/workflows/ci.yaml [hunk 2]
```diff
@@ -198,29 +198,22 @@ jobs:
...
+        name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
 
       - id: build
         name: Build Docker images
         run: |-
-          docker build --target builder \
-            --cache-from=$PREVIOUS_IMAGE_NAME-builder \
+          docker buildx build --target builder \
+            --output=type=docker \
```
Comment (alexvpickering) [inline .github/workflows/ci.yaml:219]: `--output=type=docker` required to make the images available for `docker images`, which we need for subsequent builds.


[Reference here](https://docs.docker.com/engine/reference/commandline/buildx_build/#load)

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -198,29 +198,22 @@ jobs:
...
+            --cache-from type=registry,ref=$PREVIOUS_IMAGE_NAME-builder \
             --build-arg GITHUB_PAT=$GITHUB_PAT \
             --tag $IMAGE_NAME-builder .
 
-          docker build --target prod \
+          docker buildx build --target prod \
+            --output=type=docker \
             --cache-from=$IMAGE_NAME-builder \
-            --cache-from=$PREVIOUS_IMAGE_NAME \
```
Comment (alexvpickering) [inline .github/workflows/ci.yaml:233]: we don't have the image for `$PREVIOUS_IMAGE_NAME` anymore but the `builder` target is almost the whole build so it's unnecessary.  I also don't think it ever worked (except locally).

File: .github/workflows/ci.yaml [hunk 4]
```diff
@@ -232,13 +225,16 @@ jobs:
         name: Push builder images to ECR
         run: |-
           echo Pushing image $IMAGE_NAME-builder to ECR.
-          docker push $IMAGE_NAME-builder
+          docker buildx build --push --target builder \
+            --cache-to type=registry,ref=$IMAGE_NAME-builder,mode=max,image-manifest=true \
```
Comment (alexvpickering) [inline .github/workflows/ci.yaml:239]: - `image-manifest` needed to push buildcache to ECR (see [here](https://github.com/aws/containers-roadmap/issues/876#issuecomment-1665121877))
- `mode=max` is just a safety to make sure all the layers are exported to the buildcache (not tested if necessary but have needed in other projects)

**Review by cosa65 — APPROVED** [review]

**Pull request merged**